Tasks/DownloadBuildArtifacts/main.ts

```diff
@@ -126,12 +129,30 @@ async function main(): Promise<void> {
                 // Triggering build info not found, or requested, default to specified build info
                 projectId = tl.getInput("project", true);
                 definitionId = definitionIdSpecified;
-                buildId = parseInt(tl.getInput("buildId", true));
+                buildId = parseInt(tl.getInput("buildId", artifactVersion == "specific"));
             }
         }
 
         // verify that buildId belongs to the definition selected
         if (definitionId) {
+            if (artifactVersion != "specific"){ 
+                var branchNameFilter = (artifactVersion == "latest") ? null : branchName;
+                
+                // get latest successful build filtered by branch
+                var buildsForThisDefinition = await executeWithRetries("getBuildId", () => buildApi.getBuilds( projectId, [parseInt(definitionId)],null,null,null,null,null,null,BuildStatus.Completed,BuildResult.Succeeded,null,null,null,null,null,null, BuildQueryOrder.FinishTimeDescending,branchNameFilter), 4).catch((reason) => {
+                    reject(reason);
+                    return;
+                }); 
+
+                if (!buildsForThisDefinition || buildsForThisDefinition.length == 0){ 
+                    if (artifactVersion == "latestFromBranch") reject(tl.loc("LatestBuildFromBranchNotFound", branchNameFilter));
+                    else reject(tl.loc("LatestBuildNotFound"));
+                    return;
+                }
+                console.log(tl.loc("LatestBuildFound", buildsForThisDefinition[0].id));
+                buildId = buildsForThisDefinition[0].id
+            } 
+            
```

Comment:
can we reuse the build we got in case version is latest or latestFromBranch instead of calling getBuild again in line 156?

Reply:
Yes, nice catch! Implemented & tested!